src/pudl_rmi/formatter_optimus.py

```diff
@@ -348,3 +354,69 @@ def add_balancing_account(
     logger.info(f"adding {len(balancing_records)} balancing account records.")
     ferc_deprish_eia = pd.concat([ferc_deprish_eia, balancing_records], join="outer")
     return ferc_deprish_eia
+
+
+def allocate_balancing_account_to_assets(
+    ferc_deprish_eia_w_ba: pd.DataFrame,
+    idk_ba: List[str] = [
+        "utility_name_eia",
+        "operational_status",
+        "ferc_acct_name",
+        "data_source",
+        "report_year",
+    ],
+    data_and_allocator_cols: dict = {
+        "plant_balance_w_common": ["plant_balance_w_common"],
+        "book_reserve_w_common": ["book_reserve_w_common"],
+        "depreciation_annual_epxns_w_common": ["depreciation_annual_epxns_w_common"],
+    },
+):
+    """Allocate the balancing account records across related assets.
+
+    We generate balancing account records to true up the depreciation studies
+    with the utility/FERC account level values reported to FERC1 via
+    :func:`add_balancing_account`. This function takes those balancing account
+    records and allocates them across assets. It utilizes
+    :func:`pudl_rmi.connect_deprish_to_ferc1._allocate_col`
+
+    Args:
+        ferc_deprish_eia_w_ba
+        idk_ba: list of ID columns to allocate across.
+        data_and_allocator_cols: dictionary of data columns to allocate (keys)
+            and lists of column(s) to allocate based on.
+
+    """
+    # split records
+    mask_ba = ferc_deprish_eia_w_ba.plant_name_eia == "Balancing_Account"
+    balancing_accounts = ferc_deprish_eia_w_ba.loc[mask_ba]
+    assets = ferc_deprish_eia_w_ba.loc[~mask_ba]
+    data_cols = list(data_and_allocator_cols.keys())
+    # merge the ba's onto the assets
+    assets_w_ba = pd.merge(
+        assets,
+        balancing_accounts[idk_ba + data_cols],
+        on=idk_ba,
+        suffixes=("", "_ba"),
+        how="left",
+        validate="m:1",
+    )
+    # allocate each col
+    for (data_col, allocator_cols) in data_and_allocator_cols.items():
+        assets_w_ba.loc[:, f"{data_col}_asset_level"] = _allocate_col(
+            to_allocate=assets_w_ba,
+            by=idk_ba,
+            allocate_col=f"{data_col}_ba",
+            allocator_cols=allocator_cols,
+        )
+        assets_w_ba.loc[:, f"{data_col}_w_ba"] = assets_w_ba.loc[
```

Comment:
I decided to use _w_ba as a temporary column instead of assigning directly to the data column mostly because I didn't want to overwrite the data column in case it was used as an allocator column in later data columns' allocation.... but maybe it wouldn't matter because it would be allocated proportionally?